### PR TITLE
Jenkins: Grab and dump logs to s3 on failed runs

### PR DESCRIPTION
**Pull request by bikramnehra**
## Description

- Modified `Jenkinsfile` to execute the `klog.sh` script
- Modified `s3` bucket path to accommodate `klogs`
- Modified `klog.sh` to support args from Jenkins

## Test plan

- Introduce failures in tests and trigger build
- Check whether `klog.sh` was triggered even if the build failed
- Verify if both build logs and klogs were dumped in correct s3 buckets

e.g: `https://s3.console.aws.amazon.com/s3/buckets/cap-jenkins-log/prs/PR-2170-jenkins-scf-PR-2170-6/`

### CODE REVIEW

File: Jenkinsfile [hunk 1]
```diff
@@ -870,22 +870,30 @@ pass = ${OBS_CREDENTIALS_PASSWORD}
             }
             // Save logs of failed builds to s3 - we want to analyze where we may have jenkins issues.
             script {
-                if ((env.BRANCH_NAME == 'develop' || env.BRANCH_NAME == 'master') && fileExists('bin/clean-jenkins-log')) {
+                if (env.BRANCH_NAME == 'develop' || env.BRANCH_NAME == 'master') {
                     writeFile(file: 'build.log', text: getBuildLog())
-                    sh 'bin/clean-jenkins-log'
+                    sh """
```
Comment (f0rmiga) [inline Jenkinsfile]: I think these should be 2 sh calls:
```
sh 'bin/clean-jenkins-log'
sh 'container-host-files/opt/scf/bin/klog.sh .......'
```

File: Jenkinsfile [hunk 2]
```diff
@@ -870,22 +870,30 @@ pass = ${OBS_CREDENTIALS_PASSWORD}
...
                     withAWS(region: params.S3_REGION) {
                         withCredentials([usernamePassword(
                             credentialsId: params.S3_CREDENTIALS,
                             usernameVariable: 'AWS_ACCESS_KEY_ID',
                             passwordVariable: 'AWS_SECRET_ACCESS_KEY',
                         )]) {
                             script {
-                                def subdir = "${params.S3_PREFIX}${distSubDir()}"
                                 def prefix = distPrefix()
+                                def subdir = "${params.S3_PREFIX}${distSubDir()}${prefix}${env.BUILD_TAG}/"
```
Comment (f0rmiga) [inline Jenkinsfile]: I think this is the one generating weird directory names like `PR-2170-jenkins-scf-PR-2170-6`. No need for `PR-2170` showing up twice.

File: Jenkinsfile [hunk 3]
```diff
@@ -894,4 +902,4 @@ pass = ${OBS_CREDENTIALS_PASSWORD}
             }
         }
     }
-}
+}
```
Comment (f0rmiga) [inline Jenkinsfile]: Why removing the line break?

File: container-host-files/opt/scf/bin/klog.sh [hunk 4]
```diff
@@ -21,7 +21,13 @@ if [ "$1" == "-f" ]; then
   FORCE=1
 fi
 
-NS=${1-scf}
+# read namespace val from args if provided
```
Comment (f0rmiga) [inline container-host-files/opt/scf/bin/klog.sh]: Not sure I follow this. `NS` could come either from `$1` or `$2`?

File: Jenkinsfile (hunk 1 above)
Comment (andreas-kupries) [reply Jenkinsfile]: Why ?

File: container-host-files/opt/scf/bin/klog.sh (hunk 4 above)
Comment (andreas-kupries) [reply container-host-files/opt/scf/bin/klog.sh]: Apparently yes. In Jenkins he is calling it with `klog -f NS` ...
It might be easier to handle things with `shift` to squash the processed argument(s) before going to the next.

Comment (f0rmiga) [inline container-host-files/opt/scf/bin/klog.sh]: Missing capitalization and punctuation.

Comment (andreas-kupries) [reply container-host-files/opt/scf/bin/klog.sh]: I had a log at the entire section now, and now I am confused too.
The `-f` handling does a shift. I have now no idea what this construction is for either.

Comment (f0rmiga) [reply container-host-files/opt/scf/bin/klog.sh]: Yes. https://github.com/SUSE/scf/blob/fec8cd4284c9cbc43bb0775519b2dd9599e6aded/container-host-files/opt/scf/bin/klog.sh#L20

File: Jenkinsfile (hunk 1 above)
Comment (f0rmiga) [reply Jenkinsfile]: Clearer visualization on Jenkins.

Comment (andreas-kupries) [reply Jenkinsfile]: Better separation in the log output then ?

Comment (f0rmiga) [reply Jenkinsfile]: IMO yes.

File: container-host-files/opt/scf/bin/klog.sh (hunk 4 above)
Comment (bikramnehra) [reply container-host-files/opt/scf/bin/klog.sh]: Agreed. The code seems redundant. 

`NS=${1-scf}` this bit enough

File: Jenkinsfile (hunk 1 above)
Comment (andreas-kupries) [reply Jenkinsfile]: Ok. Then agreed, two calls it should be

Comment (bikramnehra) [reply Jenkinsfile]: still not convinced how two subsequent calls will provide better logging. To my understanding, since both scripts intend log collection therefore seems logical to group them together.

File: container-host-files/opt/scf/bin/klog.sh (hunk 4 above)
Comment (bikramnehra) [reply container-host-files/opt/scf/bin/klog.sh]: now since this comment will be gone, therefore no need.

guess this comment above serves the purpose `INSTANCE_ID defaults to scf`

File: Jenkinsfile (hunk 1 above)
Comment (f0rmiga) [reply Jenkinsfile]: https://jenkins-new.howdoi.website/blue/rest/organizations/jenkins/pipelines/scf/branches/PR-2170/runs/6/nodes/81/steps/125/log/?start=0

They are grouped.

File: Jenkinsfile (hunk 2 above)
Comment (bikramnehra) [reply Jenkinsfile]: This one even seemed bit odd to me, however I kept the naming convention consistent with what was already there.

One option is to get rid of prefix which is essentially:`PR-${CHANGE_ID}-` and keep `env.BUILD_TAG`. Though this convention will diverge from rest of the script where prefix is being used

File: container-host-files/opt/scf/bin/klog.sh (hunk 4 above)
Comment (f0rmiga) [reply container-host-files/opt/scf/bin/klog.sh]: `# read namespace val from args if provided` -> `# Read namespace value from command-line arguments if provided.`

But you are right. If you'll revert this change, no need for the comment.

File: Jenkinsfile (hunk 1 above)
Comment (bikramnehra) [reply Jenkinsfile]: After looking at the logs, splitting it into two separate calls.

File: Jenkinsfile (hunk 3 above)
Comment (bikramnehra) [reply Jenkinsfile]: this is about newline at the end right?

File: Jenkinsfile (hunk 2 above)
Comment (andreas-kupries) [reply Jenkinsfile]: I am ok with this.

Comment (bikramnehra) [reply Jenkinsfile]: @andreas-kupries just to be clear, are you okay with the existing convention or my proposal?

As I haven't implemented any changes from the proposal yet for this.

File: container-host-files/opt/scf/bin/klog.sh [hunk 5]
```diff
@@ -21,7 +21,8 @@ if [ "$1" == "-f" ]; then
   FORCE=1
 fi
 
-NS=${1-scf}
+NS=${1-scf}
```
Comment (f0rmiga) [inline container-host-files/opt/scf/bin/klog.sh]: Could you revert this change? (there is an extra space at the end and a line break.

File: Jenkinsfile (hunk 2 above)
Comment (f0rmiga) [reply Jenkinsfile]: At least add a slash after `${prefix}` to make it a folder. We don't want to mess with the BUILD_TAG right now.

File: container-host-files/opt/scf/bin/klog.sh (hunk 5 above)
Comment (bikramnehra) [reply container-host-files/opt/scf/bin/klog.sh]: Done!

File: Jenkinsfile (hunk 2 above)
Comment (andreas-kupries) [reply Jenkinsfile]: I am ok with the existing convention

File: Jenkinsfile [hunk 6]
```diff
@@ -870,22 +870,28 @@ pass = ${OBS_CREDENTIALS_PASSWORD}
...
                     withAWS(region: params.S3_REGION) {
                         withCredentials([usernamePassword(
                             credentialsId: params.S3_CREDENTIALS,
                             usernameVariable: 'AWS_ACCESS_KEY_ID',
                             passwordVariable: 'AWS_SECRET_ACCESS_KEY',
                         )]) {
                             script {
-                                def subdir = "${params.S3_PREFIX}${distSubDir()}"
                                 def prefix = distPrefix()
+                                def subdir = "${params.S3_PREFIX}${distSubDir()}${prefix}${env.BUILD_TAG}/"
```
Comment (andreas-kupries) [inline Jenkinsfile]: Ok. @f0rmiga has convinced me that the current definition and path are not suitable.

The proposal is now this:

- Change "${params.S3_PREFIX}${distSubDir()}${prefix}${env.BUILD_TAG}/"
  to "${params.S3_PREFIX}${distSubDir()}${prefix}/${env.BUILD_TAG}/"
  This inserts `/` after the prefix.

- This has the problem that the path now looks like `PR-2170-/jenkins-scf-PR-2170-6`. Note the weird `-/` combination.

- To fix that remove the `-` from the output of the `distPrefix()` function. Should be around line 102 of Jenkinsfile.

- Now as `distPrefix()` is already used in other places, we have to compensate for that removal there, by adding the `-` back in where the output of `distPrefix()` is used. This is in the lines 816 (*), 821, 830. 

(*) In the line 816 it is the variable `prefix` holding the result from `distPrefix()` in line 809.

@bikramnehra ^

Comment (bikramnehra) [reply Jenkinsfile]: @andreas-kupries Agreed, however, since all other cases require `-`. Therefore, instead of making changes in the `distPrefix()`(also so many other places), we can simply omit `-` while appending to the path:

`prefix = prefix.substring(0, prefix.length() - 1)`

just before line 885

Comment (andreas-kupries) [reply Jenkinsfile]: That is definitely less work to do. Go ahead. I see no future cases at the moment.

Comment (bikramnehra) [reply Jenkinsfile]: DONE! Triggered a build to verify if everything's okay.

File: Jenkinsfile [hunk 7]
```diff
@@ -870,22 +870,29 @@ pass = ${OBS_CREDENTIALS_PASSWORD}
...
                     withAWS(region: params.S3_REGION) {
                         withCredentials([usernamePassword(
                             credentialsId: params.S3_CREDENTIALS,
                             usernameVariable: 'AWS_ACCESS_KEY_ID',
                             passwordVariable: 'AWS_SECRET_ACCESS_KEY',
                         )]) {
                             script {
-                                def subdir = "${params.S3_PREFIX}${distSubDir()}"
                                 def prefix = distPrefix()
+                                prefix = prefix.substring(0, prefix.length() - 1)
```
Comment (f0rmiga) [inline Jenkinsfile]: Last nit: put a comment stating why we are cutting the last character here, as it is not clear for anyone that didn't work on this PR (we ourselves may forget soon).

Comment (bikramnehra) [reply Jenkinsfile]: comments' in
